pravega/pravega
# Issue 3652: Segment store TLS reload upon certificate refresh

**Change log description**  
Reload TLS when segment store TLS material is refreshed. 

**Purpose of the change**  
Resolves #3652 

**What the code does**  
1. Introduces a new configuration flag in Segment Store that determines whether to reload TLS when the TLS certificate changes. 
2. The flag is off by default. If the flag is on: 
    1. Watches for change in Segment Store TLS certificate. 
        We use an event-based WatchService to watch for file modification events. That mechanism doesn't work when the file being watched is a symbolic link. Certificate management tools such as Cert Manager distribute cert as a symbolic link, so in such cases, we use a polling-based file modification monitor.
    2. Upon detecting that the certificate has changed, we replace the shared SSL context with a new SSL Context. Existing connections that were set up using the old context continue to function, while new connections use an SSL Context built from the new certificate and key.
3. The above is implemented both for distributed mode and standalone mode cluster. 
    It is supported in standalone mode to make it easier to debug and test the feature. The flag is intentionally left out of the `standalone-config.properties` file, but is a hidden flag and may be added by developers if needed. 

**How to verify it**  
All unit tests should pass. 

In addition to unit tests, I've executed the following types of tests to verify the functionality:

* Standalone mode cluster with client application writing to the cluster continuously. Replaced the certificate midway and check the client continues to function. Also, run a new client application.
* Docker-Compose cluster with the same scenario as above.
* Kubernetes cluster with Cert Manager distributing certificates to Segment Store, with the same scenario as above.

In (3), the certificate path was a symbolic link. In (1) and (2), it was a direct path (no symlink).
